### DISCUSSION TITLE
Respond with 404 on a websocket route

### DISCUSSION THREAD

**Discussion by edgarogh** [actix/actix-web#2765] [Q&A] [answered]
Hello,

I'm currently experimenting with the websocket capabilities of `actix-web`/`actix-web-actors` with a classic room-based system. I would like my websocket route to return a 404 error if a room ID is given but doesn't correspond to an existing room.

Currently, I'm doing the following at the end of an if branch, on a route handler that returns `Result<HttpResponse, actix_web::Error>`:

```rust
let err = std::io::Error::new(std::io::ErrorKind::NotFound, "room not found");
return Err(err.into());
```
_(I know that I could write cleaner code by implementing `ResponseError` but that's fine for me atm)_

When trying this branch of code, actix logs

```
DEBUG [actix_http::h1::dispatcher] cannot read request payload
DEBUG [actix_http::h1::dispatcher] handler dropped payload early; attempt to clean connection
ERROR [actix_http::h1::dispatcher] handler did not read whole payload and dispatcher could not drain read buf; return 500 and close connection
ERROR [actix_http::h1::dispatcher] stream error: Handler dropped payload before reading EOF
```

I've naively tried to consume the payload with `payload.map(|_| ()).collect::<()>().await;` but it blocks, probably because the browser expects a websocket connection to follow so it doesn't ever close its TX stream.

What can I do to return without initiating a websocket connection ?

**Comment by robjtede** [comment, marked as answer]
This was a bug. See #2779 and related issues. This might be solved now.

**Reply by edgarogh** [reply]
It is, thanks for notifying me!
